fx pass for grouping layer norm

devashishshankar opened the pull request:
Summary:
This diff adds an fx pass for grouping together `torch.nn.functional.layer_norm`. This pass is different from other pattern matcher passes, as we combine together layer_norms from different places in the graph. We're calling such fusions as "group fusions". This shows promising results for internal models.

In this diff, we also add a new type of pattern matcher pass called `GroupFusionPass`, which can be used for more types of group fusions

Differential Revision: D46124324



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @bertmaher

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+        counters["inductor"]["ln_added"] += 1
+        for i, ln in enumerate(nodes_to_fuse):
+            getitem = graph.call_function(operator.getitem, args=(group_layer_norm, i))
+            print(f"Replacing {ln} with {getitem} ({i}).")
```
Remove or replace with log

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+            torch.unbind, args=(group_layer_norm,), kwargs={"dim": stack_dim}
+        )
+
+        counters["inductor"]["ln_added"] += 1
```
full words for counter names

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+
+        for i, target_node in enumerate(target_nodes):
+            nodes_to_fuse = []
+            nx_graph = get_nx_graph_from_fx_graph(graph)
```
computing this seems expensive, can it go outside the loop?

devashishshankar replied:
So, actually no - because the graph can potentially change during each iteration of the loop. 

Initially, we tried finding all groupings and changing graph one-shot, but that didn't work as changing the graph affects future groupings.

One optimization I can do is not call this method if the graph isn't changing

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+    return G
+
+
+class GroupFusionPass:
```
Add docs

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+        stable_topological_sort(graph)
+
+
+def layer_norm_pair_check(ln1, ln2):
```
add docs

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+        and get_arg_value(ln1, 1, "normalized_shape")
+        == get_arg_value(ln2, 1, "normalized_shape")
+        and get_arg_value(ln1, 4, "eps") == get_arg_value(ln2, 4, "eps")
+    )
```
should add some sort of check that there aren't other args we are ignoring.

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,135 @@
+                self.replacement_fn(graph, nodes_to_fuse)
+
+
+def layer_norm_replacement(graph: torch.fx.Graph, nodes_to_fuse: List[torch.fx.Node]):
```
add docs

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,142 @@
+                        continue
+                    can_be_fused = self.pair_check(target_node, target_nodes[j])
+                    if (
+                        can_be_fused
```
We should add a check here that `get_mutation_region_id` is the same for the two nodes.
https://github.com/pytorch/pytorch/blob/e26f5b2ac718995822fc410fbb08ee612ab29efb/torch/_inductor/pattern_matcher.py#L799

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,188 @@
+
+
+def get_nx_graph_from_fx_graph(graph: torch.fx.Graph) -> networkx.DiGraph:
+    G = networkx.DiGraph()
```
lowercase words for variable names

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,188 @@
+        pair_check,
+        replacement_fn,
+        extra_check=lambda m: True,
+        prevent_match_across_mutations=False,
```
I worry people will forget to set this.  I think there is some could to automatically detect if we should use this.  Maybe @williamwen42 can provide a pointer.

Or we could make it required arg.

devashishshankar replied:
Made it a required arg for now

jansel replied:
This isn't a required arg (you can remove the default for that)

jansel commented on torch/_inductor/fx_passes/group_fusion.py:
```diff
@@ -0,0 +1,188 @@
+                    target_nodes.append(node)
+
+        seen_nodes = set()
+        nx_graph = get_nx_graph_from_fx_graph(graph)
```
```suggestion
        if len(target_nodes) <= 1:
            return
        nx_graph = get_nx_graph_from_fx_graph(graph)
```

To avoid compute the graph when it is not needed.

devashishshankar committed:
fx pass for grouping layer norm (#103035)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103035

This diff adds an fx pass for grouping together `torch.nn.functional.layer_norm`. This pass is different from other pattern matcher passes, as we combine together layer_norms from different places in the graph. We're calling such fusions as "group fusions". This shows promising results for internal models.

In this diff, we also add a new type of pattern matcher pass called `GroupFusionPass`, which can be used for more types of group fusions

Reviewed By: jackiexu1992

Differential Revision: D46124324

fbshipit-source-id: 9cd17210ba2310dc161d15cc525b4e6e8d945962
